### PR TITLE
[DataLoader] Add Numpy seeding to worker of DataLoader

### LINKED ISSUE #5059: np.random generates the same random numbers for each data batch
For example, the following script 

```
from __future__ import print_function

import numpy as np
import random

# np.random.seed(0)

class Transform(object):
    def __init__(self):
        pass

    def __call__(self, item = None):
        return [np.random.randint(10000, 20000), random.randint(20000,30000)]

class RandomDataset(object):

    def __init__(self):
        pass

    def __getitem__(self, ind):
        item = [ind, np.random.randint(1, 10000), random.randint(10000, 20000), 0]
        tsfm =Transform()(item)
        return np.array(item + tsfm)
    def __len__(self):
        return 20


from torch.utils.data import DataLoader

ds = RandomDataset()
ds = DataLoader(ds, 10, shuffle=False, num_workers=4)

for batch in ds:
    print(batch)
```

gives 

```
#     0   2208  10983      0  15930  26264
#     1   2798  14403      0  17685  29545
#     2    528  16195      0  12927  28761
#     3   8541  13614      0  15240  24058
#     4   7144  14373      0  18374  28081
#     5   2329  17456      0  15192  26903
#     6    423  12168      0  18504  24193
#     7   9476  12027      0  19924  22325
#     8   3427  17570      0  12895  29773
#     9   6526  13327      0  15768  24566
#[torch.LongTensor of size 10x6]
#
#
#    10   2208  15203      0  15930  26024
#    11   2798  13264      0  17685  22011
#    12    528  11714      0  12927  24688
#    13   8541  11773      0  15240  29607
#    14   7144  14655      0  18374  24573
#    15   2329  12544      0  15192  27908
#    16    423  15892      0  18504  23111
#    17   9476  17389      0  19924  23799
#    18   3427  12458      0  12895  23201
#    19   6526  14935      0  15768  20789
#[torch.LongTensor of size 10x6]
```

The second and fifth columns are np.random generated random numbers in Dataset and Transform. However, they are always that the same for different batches. This will make random data augmentation actually "not random".

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @SsnL @VitalyFedyunin @ejguan @pbelevich

### PR DESCRIPTION
**Pull request by ejguan**
Fixes #5059

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56528 [DataLoader][doc] Randomness for base_seed generator and NumPy seed
* **#56488 [DataLoader] Add Numpy seeding to worker of DataLoader**

Considering amount of requests for this feature, introduce numpy seeding as default within each worker for DataLoader.

## BC-breaking Note:
- By introducing default numpy.random seeding strategy to workers of DataLoader, users don't need to manually set seed for workers by the `worker_init_fn`. And this PR won't influence users who are currently using `worker_init_fn` to set customized seed for workers.
- DataLoader will preserve reproducibility for users who are using numpy.random within Dataset.
- Multiprocessing (without `worker_init_fn` to define seed for numpy)
  - Start method as `spawn`: Each worker will now have seed for numpy random, rather than the seed generated from the imported time of Numpy module that make the DataLoader lose the reproducibility.
  - Start method as `fork`: Each worker not only have the same benefit as `spawn`,  but also have different seed for numpy as default, rather than inheriting the same seed.

Using the following Dataset and script as an example:
```py
class RandomDataset(Dataset):
    def __getitem__(self, ind):
        item = [ind, np.random.randint(1, 10000)]
        return item

    def __len__(self):
        return 20

if __name__ == '__main__'"
    ctx = mp.get_context('fork')
    ds = RandomDataset()
    g = torch.Generator()
    g.manual_seed(0)
    dl = DataLoader(ds, 2, shuffle=False, num_workers=4, multiprocessing_context=ctx, generator=g)

    epochs = 2
    for _ in range(epochs):
        for batch in d;:
            print(batch)
        print("====" * 10)
```

### 1.8.1:
Each worker generates same random result per iteration. And the seed will be reset to same for each epoch.
```py
tensor([[   0, 7449],
        [   1, 1519]])
tensor([[   2, 7449],
        [   3, 1519]])
tensor([[   4, 9645],
        [   5, 2387]])
tensor([[   6, 9645],
        [   7, 2387]])
tensor([[   8, 3118],
        [   9, 4552]])
=========================
tensor([[   0, 7449],
        [   1, 1519]])
tensor([[   2, 7449],
        [   3, 1519]])
tensor([[   4, 9645],
        [   5, 2387]])
tensor([[   6, 9645],
        [   7, 2387]])
tensor([[   8, 3118],
        [   9, 4552]])
=========================
```

### This PR:
Each worker has different seed at the beginning and re-seed for each epoch.
```py
tensor([[   0, 8715],
        [   1, 5555]])
tensor([[   2, 6379],
        [   3, 1432]])
tensor([[   4, 3271],
        [   5, 5132]])
tensor([[   6, 4287],
        [   7, 1104]])
tensor([[   8, 8682],
        [   9, 1699]])
=========================
tensor([[   0, 1374],
        [   1,  996]])
tensor([[   2,  143],
        [   3, 3507]])
tensor([[   4, 5887],
        [   5, 4730]])
tensor([[   6, 7274],
        [   7,  738]])
tensor([[   8, 6374],
        [   9, 1572]])
=========================
```

Differential Revision: [D27908486](https://our.internmc.facebook.com/intern/diff/D27908486)

### CODE REVIEW

**Comment by ejguan** [comment]
Want to mention that I don't think we should add option to disable this Numpy seeding.
- If users have own `worker_init_fn` to set seed of numpy for each worker, this PR won't influence that.
- If users don't realize the `fork` will inherit the original seed, it's a better (or more expected) behavior for users.

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by nik-sm** [comment]
It seems like a good idea to make the default behavior "safe" - a good default behavior is very valuable. But reading through the issue that prompted this makes me feel I may be missing something.

Therefore, can I please ask a quick clarifying question: is the solution described in the docs at https://pytorch.org/docs/stable/notes/randomness.html#dataloader still correct?
It suggests to use a `worker_init_fn` like this:
```
def seed_worker(worker_id):
    worker_seed = torch.initial_seed() % 2**32
    numpy.random.seed(worker_seed)
    random.seed(worker_seed)
```

Here's a quick experiment, borrowing your example script above:

---

No `worker_init_fn` shows the problem:

```
import torch
from torch.utils.data import DataLoader, Dataset
import numpy as np
import random

class RandomDataset(Dataset):
    def __getitem__(self, ind):
        return ind, np.random.randint(1, 10000), random.randint(1, 10000)

    def __len__(self):
        return 24

dl = DataLoader(RandomDataset(), batch_size=4, shuffle=False, num_workers=6)

epochs = 2
for _ in range(epochs):
    for batch in dl:
        print(batch)
    print("====" * 10)
```

```
[tensor([0, 1, 2, 3]), tensor([2135, 5052, 7161,  109]), tensor([ 487, 9588, 4155, 9902])]
[tensor([4, 5, 6, 7]), tensor([2135, 5052, 7161,  109]), tensor([6146, 7055, 1126, 8840])]
[tensor([ 8,  9, 10, 11]), tensor([2135, 5052, 7161,  109]), tensor([1698, 4571, 1925, 2698])]
[tensor([12, 13, 14, 15]), tensor([2135, 5052, 7161,  109]), tensor([5873, 1205, 4262, 1645])]
[tensor([16, 17, 18, 19]), tensor([2135, 5052, 7161,  109]), tensor([8397, 2140, 4416, 2394])]
[tensor([20, 21, 22, 23]), tensor([2135, 5052, 7161,  109]), tensor([4236, 8497, 6571, 5048])]
========================================
[tensor([0, 1, 2, 3]), tensor([2135, 5052, 7161,  109]), tensor([5315, 6142, 5110, 6754])]
[tensor([4, 5, 6, 7]), tensor([2135, 5052, 7161,  109]), tensor([5003, 2735, 6055, 2938])]
[tensor([ 8,  9, 10, 11]), tensor([2135, 5052, 7161,  109]), tensor([6434, 7356, 6858, 6273])]
[tensor([12, 13, 14, 15]), tensor([2135, 5052, 7161,  109]), tensor([3947, 7260, 2670, 4089])]
[tensor([16, 17, 18, 19]), tensor([2135, 5052, 7161,  109]), tensor([9127, 8769, 5620, 9274])]
[tensor([20, 21, 22, 23]), tensor([2135, 5052, 7161,  109]), tensor([3794, 2942, 6791, 4965])]
========================================
```

---

Adding `worker_init_fn` as described in the docs fixes problem **(unless I am missing something?):**

```
import torch
from torch.utils.data import DataLoader, Dataset
import numpy as np
import random

class RandomDataset(Dataset):
    def __getitem__(self, ind):
        return ind, np.random.randint(1, 10000), random.randint(1, 10000)

    def __len__(self):
        return 24

def wif(worker_id):
    worker_seed = torch.initial_seed() % 2**32
    np.random.seed(worker_seed)
    random.seed(worker_seed)

dl = DataLoader(RandomDataset(), batch_size=4, shuffle=False, num_workers=4, worker_init_fn=wif)

epochs = 2
for _ in range(epochs):
    for batch in dl:
        print(batch)
    print("====" * 10)
```

```
[tensor([0, 1, 2, 3]), tensor([9577, 9408, 2774, 9216]), tensor([2722, 8934, 1724,  332])]
[tensor([4, 5, 6, 7]), tensor([1837, 3923, 6292, 9825]), tensor([6932, 7012, 6218, 8936])]
[tensor([ 8,  9, 10, 11]), tensor([5965,  477, 3511, 8001]), tensor([3448, 2379, 9089, 7660])]
[tensor([12, 13, 14, 15]), tensor([1204,   20, 2110, 2399]), tensor([8398,  478,   80, 2960])]
[tensor([16, 17, 18, 19]), tensor([8268, 7027,  616, 2815]), tensor([5523, 5037, 1965, 8266])]
[tensor([20, 21, 22, 23]), tensor([4388, 8424, 3568, 3933]), tensor([2505,  715, 2573,  371])]
========================================
[tensor([0, 1, 2, 3]), tensor([8847, 6474, 1756, 7415]), tensor([4453, 4118, 1848, 2441])]
[tensor([4, 5, 6, 7]), tensor([9985, 3272, 5960, 8155]), tensor([6547, 9068, 9686, 5893])]
[tensor([ 8,  9, 10, 11]), tensor([5527, 8779, 2296, 2978]), tensor([7890, 9320, 6835, 9731])]
[tensor([12, 13, 14, 15]), tensor([5010, 7736, 8084, 4132]), tensor([2160, 8856, 5288, 9725])]
[tensor([16, 17, 18, 19]), tensor([ 681, 4222, 9951, 4279]), tensor([1339, 9125, 9191, 3671])]
[tensor([20, 21, 22, 23]), tensor([2913, 7189, 9442, 4418]), tensor([9893, 4789, 8781,  722])]
========================================
```

Thanks!

**Comment by ejguan** [comment]
@nik-sm 
As you can see in your first experiment, before this PR, the NumPy random value for each worker remains exactly same for each iteration.
Now, if you test this script against this PR on master branch, the NumPy random value would vary for each (batch) from each worker.

And, it also works with the `worker_init_fn` in the document from https://pytorch.org/docs/stable/notes/randomness.html#dataloader. And, by using the `worker_init_fn`, the seed you set in the `worker_init_fn` will override the default seed.

File: torch/utils/data/_utils/worker.py [hunk 1]
```diff
@@ -134,8 +134,13 @@ def _worker_loop(dataset_kind, dataset, index_queue, data_queue, done_event,
         signal_handling._set_worker_signal_handlers()
 
         torch.set_num_threads(1)
+        seed = base_seed + worker_id
         random.seed(seed)
         torch.manual_seed(seed)
+        if HAS_NUMPY:
+            import numpy as np
+            ss = np.random.SeedSequence([worker_id, base_seed])
```
Comment (rgommers) [inline torch/utils/data/_utils/worker.py:142]: @ejguan you're dropping compatibility with `numpy < 1.17.0` here. That probably wasn't intended. This is _just_ okay for conda-forge, but in the `builder` repo lower versions of numpy specified.

Since all you need is an array of four `uint32` numbers, I recommend checking `np.__version__` and if it's lower than `'1.17'` provide a fallback.

Comment (rkern) [reply torch/utils/data/_utils/worker.py:142]: Just be careful with the fallback. Don't reuse the same seed for the stdlib `random` and `np.random`, for example. [Bad things happen.](https://github.com/PyTorchLightning/pytorch-lightning/pull/6960#issuecomment-818393659)

I wouldn't say that "all you need is _an_ array of `uint32` numbers". Safely deriving parallel streams is a tricky business. The PRNGs in play here (MT19937 and MTGP) just don't have the properties that we would want them to in order to use simple seed derivation schemes. What particular fallback were you thinking of?

Comment (ejguan) [reply torch/utils/data/_utils/worker.py:142]: @rgommers Oh, thank you for bringing it up. Will have a patch to fix it.

Comment (ejguan) [reply torch/utils/data/_utils/worker.py:142]: @rkern Thanks for providing insights on seeding for RNG. Just wondering if it falls back to an array of four `uint32` number, which way do you suggest to generate these four numbers?

Comment (rkern) [reply torch/utils/data/_utils/worker.py:142]: Honestly, I suggest vendorizing `SeedSequence` (or its [C++ progenitor](https://gist.github.com/imneme/540829265469e673d045)).

Comment (ejguan) [reply torch/utils/data/_utils/worker.py:142]: I am thinking, at least for now, to utilize the 64bit int from torch seed for numpy to have a better random performance

Comment (rkern) [reply torch/utils/data/_utils/worker.py:142]: Ah, that was the [Bad things happen](https://github.com/PyTorchLightning/pytorch-lightning/pull/6960#issuecomment-818393659) option that I warn against. :-)

Both the stdlib `random` and `np.random` (pre 1.17) both use the same MT19937 PRNG algorithm, and their seeding algorithms are the same (largely). You do not want them both to start with the same state because both might get used in the worker (I see a lot of code that mixes `random` and `np.random` in ML workflows). You want both of those PRNG streams to be decorrelated from each other. The integer hashes underlying `SeedSequence` give you a good way to deterministically derive high-quality sets of seeds derived from sets that are convenient (e.g. `base_seed+i`) but merely distinct. Distinctness isn't _quite_ enough for MT19937 the same way it is for Philox. So, for example, I also wouldn't try to use 

```python
random.seed(torch_seed)
np.random.seed([torch_seed % (1<<32), torch_seed >> 32, SOME_CONSTANT_TO_MAKE_THEM_DISTINCT])
```

The MT19937 seeding algorithm is _okay_, but it wasn't really designed to be stress-tested in this way to generate parallel streams. `SeedSequence` was.

Comment (ejguan) [reply torch/utils/data/_utils/worker.py:142]: Understood, same seed leads to same result with same algorithm. And, you are right. I think the ultimate goal or solution is to have `SeedSequence` implemented in PyTorch. But, before that, I can only provide a solution by making sure the distinctness, but with better seeding algorithm and randomness for `numpy.random` and `random` module. Does that make sense?

Comment (rkern) [reply torch/utils/data/_utils/worker.py:142]: I'm not sure what's stopping you from vendorizing `SeedSequence` now.

But if you need a quick stopgap, you can pass the 64-bit seed through [SplitMix64](https://xoshiro.di.unimi.it/splitmix64.c) or some other similar [64-bit integer hash](https://nullprogram.com/blog/2018/07/31/) (there are a lot that use this pattern, but do make sure to use the initial addition that SplitMix64 does to prevent 0 from mapping to 0).

```python
M64 = 0xFFFFFFFF_FFFFFFFF
def splitmix64(u64):
    z = (u64 + 0x9e3779b97f4a7c15) & M64
    z = (z ^ (z >> 30)) & M64
    z = (z * 0xbf58476d1ce4e5b9) & M64
    z = (z ^ (z >> 27)) & M64
    z = (z * 0x94d049bb133111eb) & M64
    z = (z ^ (z >> 31)) & M64
    return z

torch_seed = torch.initial_seed()
# We can reuse the PyTorch's seed once because it uses Philox and random uses MT19937.
random.seed(torch_seed)
# But we can't reuse it a second time for np.random since it also uses MT19937.
# Permute it deterministically to ensure that it is "far" from torch_seed.
numpy_seed = splitmix64(torch_seed)
np.random.seed([numpy_seed % (1<<32), numpy_seed >> 32])
```

Comment (ejguan) [reply torch/utils/data/_utils/worker.py:142]: We will utilize SeedSequence for NumPy >= 1.17.0, but it's not available for prior NumPy releases

Comment (rkern) [reply torch/utils/data/_utils/worker.py:142]: Right, that's why I recommended vendorizing it.

Comment (ejguan) [reply torch/utils/data/_utils/worker.py:142]: I have no objection to have SeedSequence. But, I prefer fixing it as it's a bug.

Comment (rkern) [reply torch/utils/data/_utils/worker.py:142]: Vendorizing `SeedSequence` (i.e. copying the [`SeedSequence` implementation](https://github.com/numpy/numpy/blob/main/numpy/random/bit_generator.pyx#L252-L476) into PyTorch or reimplementing the algorithm in whatever way is convenient for you) would fix the bug. Importantly, it would fix it the same way for all versions of numpy and not have two different behaviors depending on which version of numpy the user had installed.

Comment (rgommers) [reply torch/utils/data/_utils/worker.py:142]: I was thinking to derive the numbers via a simple scheme from the pytorch PRNG indeed. However, I agree that vendoring `SeedSequence` is better. Should be the C++ progenitor you linked to above, because PyTorch doesn't have a build dependency on Cython, and that would be annoying to add. Or just translate the `SeedSequence` implementation to pure Python; performance for a one-off call is going to be fine compared to loading lots of data.

Comment (rkern) [reply torch/utils/data/_utils/worker.py:142]: Yeah, the only wrinkle to implementing `SeedSequence` in Python is that you have to be explicit about enforcing the `uint32` clock arithmetic.

Comment (ejguan) [reply torch/utils/data/_utils/worker.py:142]: One PR #56797 is submitted with a hard-code state generation for NumPy. And an issue #56805 is also submitted for feature request of `SeedSequence`. Feel free to review and comment on them. Thanks a lot!

**Comment by xychen2022** [comment]
Hello,

I have a problem with initializing different workers with different seeds. It totally makes sense to me that different workers should load different data, but my experiments show that this leads to worse performance. Have any of you came across the same problem? And how did you solve it? Thanks!
